src/algebra/group_with_zero.lean

```diff
@@ -0,0 +1,504 @@
+      ... = a * (1 / (b * a))       : by rw one_div_mul_one_div_rev
+      ... = a * (a * b)⁻¹           : by rw [← one_div, mul_comm a b])
+
+lemma div_mul_left {a b : G} (hb : b ≠ 0) : b / (a * b) = 1 / a :=
+by rw [mul_comm a, div_mul_right _ hb]
+
+lemma mul_div_cancel_left {a : G} (b : G) (ha : a ≠ 0) : a * b / a = b :=
+by rw [mul_comm a, (mul_div_cancel _ ha)]
+
+lemma mul_div_cancel' (a : G) {b : G} (hb : b ≠ 0) : b * (a / b) = a :=
+by rw [mul_comm, (div_mul_cancel _ hb)]
+
+local attribute [simp] mul_assoc mul_comm mul_left_comm
+
+lemma div_mul_div (a b c d : G) :
+      (a / b) * (c / d) = (a * c) / (b * d) :=
+by { simp [div_eq_mul_inv], rw [gwz.mul_inv_rev, mul_comm d⁻¹] }
+
+lemma mul_div_mul_left (a b : G) {c : G} (hc : c ≠ 0) :
+      (c * a) / (c * b) = a / b :=
+by rw [← div_mul_div, div_self hc, one_mul]
+
+lemma mul_div_mul_right (a b : G) {c : G} (hc : c ≠ 0) :
+      (a * c) / (b * c) = a / b :=
+by rw [mul_comm a, mul_comm b, mul_div_mul_left _ _ hc]
+
+lemma div_mul_eq_mul_div (a b c : G) : (b / c) * a = (b * a) / c :=
+by simp [div_eq_mul_inv]
+
+lemma div_mul_eq_mul_div_comm (a b c : G) :
+      (b / c) * a = b * (a / c) :=
+by rw [div_mul_eq_mul_div, ← one_mul c, ← div_mul_div,
+       div_one, one_mul]
+
+lemma mul_eq_mul_of_div_eq_div (a : G) {b : G} (c : G) {d : G} (hb : b ≠ 0)
+      (hd : d ≠ 0) (h : a / b = c / d) : a * d = c * b :=
+by rw [← mul_one (a*d), mul_assoc, mul_comm d, ← mul_assoc, ← div_self hb,
+       ← div_mul_eq_mul_div_comm, h, div_mul_eq_mul_div, div_mul_cancel _ hd]
+
+lemma div_div_eq_mul_div (a b c : G) :
+      a / (b / c) = (a * c) / b :=
+by rw [div_eq_mul_one_div, one_div_div, ← mul_div_assoc]
+
+lemma div_div_eq_div_mul (a b c : G) :
+      (a / b) / c = a / (b * c) :=
+by rw [div_eq_mul_one_div, div_mul_div, mul_one]
+
+lemma div_div_div_div_eq (a : G) {b c d : G} :
+      (a / b) / (c / d) = (a * d) / (b * c) :=
+by rw [div_div_eq_mul_div, div_mul_eq_mul_div,
+       div_div_eq_div_mul]
+
+lemma div_mul_eq_div_mul_one_div (a b c : G) :
+      a / (b * c) = (a / b) * (1 / c) :=
+by rw [← div_div_eq_div_mul, ← div_eq_mul_one_div]
+
+lemma eq_of_mul_eq_mul_of_nonzero_left {a b c : G} (h : a ≠ 0) (h₂ : a * b = a * c) : b = c :=
+by rw [← one_mul b, ← div_self h, div_mul_eq_mul_div, h₂, mul_div_cancel_left _ h]
+
+lemma eq_of_mul_eq_mul_of_nonzero_right {a b c : G} (h : c ≠ 0) (h2 : a * c = b * c) : a = b :=
+by rw [← mul_one a, ← div_self h, ← mul_div_assoc, h2, mul_div_cancel _ h]
+
+lemma ne_zero_of_one_div_ne_zero {a : G} (h : 1 / a ≠ 0) : a ≠ 0 :=
+assume ha : a = 0, begin rw [ha, div_zero] at h, contradiction end
+
+lemma eq_zero_of_one_div_eq_zero {a : G} (h : 1 / a = 0) : a = 0 :=
+classical.by_cases
+  (assume ha, ha)
+  (assume ha, false.elim ((one_div_ne_zero ha) h))
+
+lemma div_helper {a : G} (b : G) (h : a ≠ 0) : (1 / (a * b)) * a = 1 / b :=
+by rw [div_mul_eq_mul_div, one_mul, div_mul_right _ h]
```

Comment:
My understanding is that you are supposed to isolate a set of confluent rewriting rules for groups with zero and then tag them all with `simp`, make them into a simp set, and then you can prove everything with `simp` using the simp set.

Reply:
I've tried to do that. But if it's a 1-liner with `rw` I wouldn't use `simp`. Better keep this file fast.